### DISCUSSION TITLE
Import & Export glb at runtime

### DISCUSSION THREAD

**Discussion by SadiAtBim** [atteneder/glTFast#678] [General]
I have some use cases to include in this discussion. Please help me to find a solution for my requirements.

1 Using gltFast package, i am successfully downloading glb (skinnedmeshes) file from the server and applying on my avatar perfectl. All of my parts of skin of avatars are applied with the bones i have. And all parts are perfectly working while animation running.

2 Now the second point is, I am exporting customized avatar which have some runtime glb parts applied on it. 
I used SimpleExport and AdvanceExport methods to export this avatar as GLB. BUT these two methods are not adding textures on exported GLB nor adding blendshapes data if any part has on it.
`
// here is the methods
    public async void AdvancedExport(GameObject avatar, string fileName)
    {

        string path = "some_path" + fileName;
        if (!Directory.Exists(path))
            Directory.CreateDirectory(path);
        // CollectingLogger lets you programmatically go through
        // errors and warnings the export raised
        var logger = new CollectingLogger();

        // ExportSettings and GameObjectExportSettings allow you to configure the export
        // Check their respective source for details

        // ExportSettings provides generic export settings
        var exportSettings = new ExportSettings
        {
            Format = GltfFormat.Binary,
            FileConflictResolution = FileConflictResolution.Overwrite,
            ImageDestination = ImageDestination.MainBuffer,
            // Export everything except cameras or animation
            ComponentMask = ~(ComponentType.Camera | ComponentType.Animation),
            // Boost light intensities 
            LightIntensityFactor = 100f,

            //// Enable Draco compression
            //Compression = Compression.Draco,
            //// Tweak the Draco compression settings
            //DracoSettings = new DracoExportSettings
            //{
            //    positionQuantization = 12
            //}

        };

        // GameObjectExportSettings provides settings specific to a GameObject/Component based hierarchy
        var gameObjectExportSettings = new GameObjectExportSettings
        {
            // Include inactive GameObjects in export
            OnlyActiveInHierarchy = true,
            // Also export disabled components
            DisabledComponents = false,
            // Only export GameObjects on certain layers
            LayerMask = LayerMask.GetMask("Default"),

        };

        // GameObjectExport lets you create glTFs from GameObject hierarchies
        var export = new GameObjectExport(exportSettings, gameObjectExportSettings, logger: logger);

        // Example of gathering GameObjects to be exported (recursively)
        var rootLevelNodes = new[] { avatar };

        // Add a scene
        export.AddScene(rootLevelNodes, "My new glTF scene");

        // Async glTF export
        var success = await export.SaveToFileAndDispose(path + $"/{fileName}.glb");

        if (!success)
        {
            UnityEngine.Debug.LogError("Something went wrong exporting a glTF");
            // Log all exporter messages
            logger.LogAll();
        }
    }
    public async void SimpleExport(GameObject targetGameObject, string fileName)
    {
        string path = "some_path" + fileName;
        if (!Directory.Exists(path))
            Directory.CreateDirectory(path);

        // Example of gathering GameObjects to be exported (recursively)
        var rootLevelNodes = new[] { targetGameObject };

        // GameObjectExport lets you create glTFs from GameObject hierarchies
        var export = new GameObjectExport();

        // Add a scene
        export.AddScene(rootLevelNodes);

        // Async glTF export
        bool success = await export.SaveToFileAndDispose(path + $"/{fileName}.glb");
        //bool success = await export.SaveToStreamAndDispose(path);

        if (!success)
        {
            UnityEngine.Debug.LogError("Something went wrong exporting a glTF");
        }
    }`

Here is the result of point 2:
![image](https://github.com/atteneder/glTFast/assets/129041312/d31b6e85-0ea8-4ed4-971a-0437585fb012)
 

3 Due to false result, i decided to use UnityGLTF exporting logics into my project. I integrated the logics of UnityGLTF for exporting customized avatar. It has given me very good results. It has applied textures and materials. It also added the blendshapes in some skinnedmeshes which has blendshapes.

`
// here is the methods
    internal void ExportGLB(GameObject gameObject, string fileName)
    {

        //if (simpleExport) SimpleExport(gameObject, path, fileName);
        //else AdvancedExport(gameObject, path, fileName);

        string path = AvatarController.Instance.AvatarBuilderConfig.exportAvatarPath + fileName;
        if (!Directory.Exists(path))
            Directory.CreateDirectory(path);

        //gameObject.GetComponentInChildren<Animator>().enabled = true;

        //Camera.main.cullingMask = -1;
        gameObject.transform.rotation = Quaternion.identity;

        GLTFSettings settings = Resources.Load("UnityGLTFSettings") as GLTFSettings;
        GLTFSceneExporter exporter;
        if (settings != null)
        {
            UnityEngine.Debug.Log("Exporting with settings");
            exporter = new GLTFSceneExporter(new[] { gameObject.transform }, new ExportOptions(settings));
        }
        else
        {
            UnityEngine.Debug.Log("Exporting with default settings");
            exporter = new GLTFSceneExporter(new[] { gameObject.transform }, new ExportOptions());
        }

        //ChangeShadertoPBR(exporter, avatarGameobject);

        byte[] recievedBytes = exporter.SaveGLBToByteArray(gameObject.name);

        //Camera.main.cullingMask = 1 << 5;

        if (recievedBytes == null)
        {
            UnityEngine.Debug.Log("Failed to convert into bytes");

        }
        else
        {
            UnityEngine.Debug.Log("GLB Bytes " + recievedBytes.Length);
            SaveGLBToFile(recievedBytes, path, fileName + ".glb");

        }
    }
`

Here is the result of point 3:
![image](https://github.com/atteneder/glTFast/assets/129041312/432bb53b-9079-4ead-aaff-5784841a8171)

So difference is: 
 
![image](https://github.com/atteneder/glTFast/assets/129041312/8250b416-7cdb-445c-ba59-9496da00df5b)

The exported GLB by using point 3 is good to go if we use the glb in unity. but if we import into Unreal Engine 5. It shows the following errors and in some exported with blendshapes data crashes UE5. 
![image](https://github.com/atteneder/glTFast/assets/129041312/774e69ca-6ffa-4af4-b64d-e798393ea251)

Question 1: Runtime importing with gltFast package is very good, but exporting at runtime is not perfect. How to use glb exporting at runtime perfectly? Will the exported GLB working on unreal and other engine as standard glb ?

Question 2: UnityGLTF (Khronos) is giving very good exporting at runtime but when the blendshapes comes in, it export the glb but its  not working in UE (working in Unity only). If we view it [online ](https://gltf-viewer.donmccurdy.com/)it shoes errors. it also crashes UE while importing this glb. What to do with export?

I just wanted to import glbs at runtime any skinnedmeshes either its has blendshapes or not. After this, i only wanted to export the game object having multiple skinnedmeshes applied at runtime with good texturing and materials as well as it should add blendshape perfectly, The last i wanted my exported GLB must be useable in any platform or engine or tools without having problem in it. 

Please guide me if anyone have the solution for my requirements. Thanks.

**Comment by atteneder** [comment]
Hi,

the limitation you're facing is known (see #351). A colleague of mine has a solution in active development. I plan to have a look at it no sooner than in 2 to 4 weeks.

Regarding question 1:
The export is automatically tested, including a glTF validator check, so it's unlikely that glTFast produces invalid files. Of course, there's always exceptions, but nothing that cannot be fixed.

Regarding question 2:
I cannot offer support for UnityGLTF.

hth

**Reply by SadiAtBim** [reply]
@atteneder in my project i am downloading the glb at runtime. Is there any issue with it? if the asset is downloaded from gltf asset then it will if exported at runtime to save as glb file. Exporting code is not including the textures in exported glb. but materials are assigned. what possible mistake i am doing in export side?

**Reply by atteneder** [reply]
You likely did nothing wrong. Support for exporting glTFast's own materials including textures was added in version 6.5.0, so you should try to update to the most recent version and try it again.
